onnxruntime/core/common/threadpool.cc

```diff
@@ -427,19 +425,22 @@ void ThreadPool::ParallelForFixedBlockSizeScheduling(const std::ptrdiff_t total,
     RunInParallel(run_work, num_work_items, block_size);
   } else {
     int num_of_blocks = d_of_p * thread_options_.dynamic_block_base_;
-    std::ptrdiff_t block_size_base{1};
-    while (block_size_base * num_of_blocks < total) block_size_base <<= 1;
-    std::atomic<ptrdiff_t> iter{0};
-    std::function<void(unsigned)> run_work = [&](unsigned) {
-      std::ptrdiff_t start{0};
-      std::ptrdiff_t b = block_size_base;
-      while ((start = iter.fetch_add(b, std::memory_order_relaxed)) < total) {
-        fn(start, std::min(start + b, total));
-        auto residual = total - start;
-        while (b > 1 && (b * num_of_blocks > residual)) b >>= 1;
+    std::ptrdiff_t base_block_size = std::max(1LL, std::llroundl(static_cast<long double>(total) / num_of_blocks));
+    alignas(CACHE_LINE_BYTES) std::atomic<ptrdiff_t> left{total};
+    LoopCounter lc(total, d_of_p, base_block_size);
+    std::function<void(unsigned)> run_work = [&](unsigned idx) {
+      std::ptrdiff_t b = base_block_size;
+      unsigned my_home_shard = lc.GetHomeShard(idx);
+      unsigned my_shard = my_home_shard;
+      uint64_t my_iter_start, my_iter_end;
+      while (lc.ClaimIterations(my_home_shard, my_shard, my_iter_start, my_iter_end, b)) {
+        fn(static_cast<std::ptrdiff_t>(my_iter_start),
+           static_cast<std::ptrdiff_t>(my_iter_end));
+        auto todo = left.fetch_sub(my_iter_end - my_iter_start, std::memory_order_relaxed);
+        if (b > 1) b = std::max(1LL, std::llroundl(static_cast<long double>(todo) / num_of_blocks));
```

Comment:
```suggestion
        // Re-compute the amount of work remaining, and adjust the
        // block size if appropriate.
        // TODO : If we make dynamic sizing the default, then we should
        // investigate techniques to avoid an atomic here.  For instance, we
        // could step down the next block size algorithmically based on the
        // amount of work that we have claimed.
        auto todo = left.fetch_sub(my_iter_end - my_iter_start, std::memory_order_relaxed);
        if (b > 1) b = std::max(1LL, std::llroundl(static_cast<long double>(todo) / num_of_blocks));
```

Reply:
Done.